### PR TITLE
Implement AddHunt UI

### PR DESCRIPTION
**Pull request by BuseThai**
## **User Story**
As a user, I want to create a new hunt with its main details in order to share it with others. 

---

## **Task Reference**
**Task:** Implement and test `Add Hunt` screens and map selection flow  
**Related Feature:** Hunt Creation UI Flow
**Related Issue:** https://github.com/SwentSeekr/seekr/issues/42

---

## **Description**
This PR introduces the **complete Add Hunt UI flow**, allowing users to input hunt details, select map points, and save their hunt.  

It adds three new composable screens, consistent test tags, and a comprehensive **instrumentation test suite** covering all interactive and validation behaviors.

---

###  Core Implementation
- **New Screens Added**
  - `AddHuntScreen.kt`  
    - Coordinates between form and map screens  
    - Shows toast messages for errors and validations  
    - Handles navigation between input and map selection  
  - `AddHuntFieldsScreen.kt`  
    - UI for entering title, description, estimated time, distance, difficulty, and status  
    - Real-time validation and error feedback  
    - “Select Locations” button dynamically updates with point count  
    - “Save Hunt” button enabled only when form is valid  
  - `AddPointsMapScreen.kt`  
    - Google Map interface for selecting hunt points  
    - Displays markers and connecting polylines  
    - “Confirm Points” and “Cancel” buttons with dynamic state  
    - 
---

## **Demo video**

https://github.com/user-attachments/assets/7665420b-ec19-4b6e-9713-200a585aa58a

---

## **✅ Checklist**
- [x] Implemented `AddHuntFieldsScreen`, `AddHuntScreen`, and `AddPointsMapScreen`  
- [x] Added test tags for UI automation  
- [x] Implemented input validation and dynamic button labels  
- [x] Added instrumentation tests for all UI states  
- [x] Verified all tests pass locally and in CI  
- [x] Linked with existing `AddHuntViewModel` logic  

---

## **Files Changed**

| File | Description |
|------|--------------|
| `app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt` | Root composable coordinating fields and map selection |
| `app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreen.kt` | UI for entering hunt details with validation |
| `app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreen.kt` | Map selection screen for hunt points |
| `app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreenTest.kt` | Tests validation, dropdowns, and save logic |
| `app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreenTest.kt` | Tests integration and navigation |
| `app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreenTest.kt` | Tests map selection, confirm, and cancel logic |

### CODE REVIEW

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    addHuntViewModel: AddHuntViewModel = viewModel(),
+    onGoBack: () -> Unit = {},
+    onDone: () -> Unit = {}
+) {
+  var isSelectingPoints by remember { mutableStateOf(false) }
+  val uiState by addHuntViewModel.uiState.collectAsState()
+  val context = LocalContext.current
+
+  // Handle error toast
+  LaunchedEffect(uiState.errorMsg) {
```
Comment (nolanprad) [inline app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt]: Toast-based error handling is not testable, toasts can’t be asserted in instrumentation tests and aren’t accessible.

Fix: Replace with a Snackbar via Scaffold to show transient messages; it’s testable and accessible.
Alternatively, wrap the toast in a showMessage callback injected via parameter for testing.

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      addHuntViewModel.clearErrorMsg()
+    }
+  }
+
+  if (isSelectingPoints) {
+    AddPointsMapScreen(
+        initPoints = uiState.points,
+        onDone = { locations ->
+          val points = locations.toMutableList()
+          if (points.size < 2) {
```
Comment (nolanprad) [inline app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt]: Map validation logic in composable breaks architecture.
Problem:
Business validation (“must have 2+ points”) is UI-layer logic.
This violates separation of concerns — should be in the ViewModel.
Fix:
Move this validation inside addHuntViewModel.setPoints() or a dedicated validatePoints() method.
Emit a uiState.errorMsg or typed event if invalid.
The UI just reacts.

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    AddHuntFieldsScreen(
+        uiState = uiState,
+        onTitleChange = addHuntViewModel::setTitle,
+        onDescriptionChange = addHuntViewModel::setDescription,
+        onTimeChange = addHuntViewModel::setTime,
+        onDistanceChange = addHuntViewModel::setDistance,
+        onDifficultySelect = addHuntViewModel::setDifficulty,
+        onStatusSelect = addHuntViewModel::setStatus,
+        onSelectLocations = { isSelectingPoints = true },
+        onSave = { if (addHuntViewModel.addHunt()) onDone() },
```
Comment (nolanprad) [inline app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt]: addHuntViewModel.addHunt() return type misuse
Problem:
If addHunt() is suspend or emits success asynchronously, this won’t work.
UI may navigate before the ViewModel confirms success.
Fix:
Make addHunt() update a state (e.g., uiState.saveSuccess).
Observe via LaunchedEffect(uiState.saveSuccess) → then call onDone().
This avoids navigation race conditions.

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt
Comment (nolanprad) [inline app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt]: The file is clean, readable, and effectively coordinates the Add Hunt flow, but it mixes UI logic with business validation and state management that belong in the ViewModel. Overall, it’s a solid foundation that just needs better architectural separation and testable feedback handling.
Missing lifecycle-safe navigation
When user rotates the device mid-way (e.g., while selecting points), the flow resets to AddHuntFieldsScreen since isSelectingPoints is not saveable.
Fix: Use rememberSaveable (if kept local) or persist via ViewModel

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreen.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun AddPointsMapScreen(
+    onDone: (List<Location>) -> Unit,
+    initPoints: List<Location> = emptyList(),
+    onCancel: () -> Unit
+) {
+  var points by remember { mutableStateOf(initPoints) }
```
Comment (nolanprad) [inline app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreen.kt]: Local point state should be ViewModel-driven
Problem: Point state is ephemeral and lost on configuration change or process death.
Fix:
Use rememberSaveable (if you want quick local persistence), or
Manage points entirely in the ViewModel (uiState.points) and make this composable stateless.
This ensures data consistency when navigating between screens

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreen.kt
Comment (nolanprad) [inline app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreen.kt]: This screen is visually and structurally solid, showing good Compose practices.
It mainly needs architectural cleanup (state hoisting, validation relocation) and a few UX/accessibility.

Missing lifecycle handling for camera, no camera update or focus when new points are added; camera may remain static far from markers.
After each addition, optionally animate camera to the last added point:
LaunchedEffect(points.lastOrNull()) {
    points.lastOrNull()?.let {
        cameraPositionState.animate(CameraUpdateFactory.newLatLng(LatLng(it.latitude, it.longitude)))
    }
}

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreen.kt
Comment (nolanprad) [inline app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreen.kt]: This is a well-structured and readable form composable that follows Compose conventions nicely.
Its main shortcomings are mostly UI-state repetition and hardcoded text, but functionally it’s robust and close to production quality.

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreenTest.kt [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      onDone: (List<Location>) -> Unit = {},
+      onCancel: () -> Unit = {}
+  ) {
+    composeRule.setContent {
+      MaterialTheme {
+        AddPointsMapScreen(onDone = onDone, initPoints = initPoints, onCancel = onCancel)
+      }
+    }
+  }
+
```
Comment (nolanprad) [inline app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreenTest.kt]: Missing test for dynamic map interaction, like:
@Test
fun clickingMap_addsNewPoints_andEnablesConfirm() {
    setContent()
    // simulate adding two points
    composeRule.runOnUiThread {
        // Trigger map callback manually since GoogleMap isn't interactive in Compose tests
        fakeAddPoint(Location(0.0, 0.0, "Point 1"))
        fakeAddPoint(Location(1.0, 1.0, "Point 2"))
    }
    composeRule.onNodeWithTag(AddPointsMapScreenTestTags.CONFIRM_BUTTON)
        .assertIsEnabled()
        .assertTextContains("Confirm Points (2)")
}

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreenTest.kt
Comment (nolanprad) [inline app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreenTest.kt]: These tests are concise, readable, and well-structured, covering the most basic behaviors correctly.
However, they currently verify only static UI states, not actual map interactions or dynamic state changes, which are essential to fully validate this screen’s functionality.
Could extend the second test to verify that markers (if tagged or countable in future versions) match the number of points, and add a test verifying that pressing cancel after adding points properly calls onCancel() and doesn’t trigger onDone().

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreenTest.kt [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  @Test
+  fun starts_with_fields_and_back_navigates() {
+    val vm = AddHuntViewModel()
+    var backCalled = false
+    setContent(vm, onGoBack = { backCalled = true })
+
+    composeRule.onNodeWithText("Add your Hunt").assertExists()
+    composeRule.onNodeWithTag(AddHuntScreenTestTags.BUTTON_SELECT_LOCATION).assertExists()
```
Comment (nolanprad) [inline app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreenTest.kt]: Hardcoded UI strings will break with localization, could use test tags for stable selectors:
composeRule.onNodeWithTag(AddHuntScreenTestTags.TITLE).assertExists()
or inject localized string via stringResource(R.string.add_your_hunt) in the test setup.

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreenTest.kt
Comment (nolanprad) [inline app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreenTest.kt]: This test class is strongly structured and functionally meaningful, covering the main user interactions end-to-end.
Its biggest gap is over-reliance on the real ViewModel and missing happy-path coverage, but once those are addressed, it’ll provide robust, maintainable validation for the entire Add Hunt flow.
Missing success flow test:
@Test
fun valid_state_logged_in_triggers_onDone() {
    val vm = FakeAddHuntViewModelReturningSuccess()
    var done = false
    setContent(vm, onDone = { done = true })
    // Fill form fields...
    composeRule.onNodeWithTag(AddHuntScreenTestTags.HUNT_SAVE).performClick()
    composeRule.waitForIdle()
    assertTrue(done)
}

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreenTest.kt [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        )
+      }
+    }
+  }
+
+  @Test
+  fun renders_topBar_and_backButton_triggers() {
+    setContent()
+
+    composeRule.onNodeWithText("Add your Hunt").assertExists()
```
Comment (nolanprad) [inline app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreenTest.kt]: Hardcoded UI strings in assertions, should composeRule.onNodeWithTag(AddHuntScreenTestTags.TITLE).assertExists()
composeRule.onNodeWithTag(AddHuntScreenTestTags.BACK_BUTTON).performClick()

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreenTest.kt [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    // Open and pick first difficulty
+    val diffField = composeRule.onNodeWithTag(AddHuntScreenTestTags.DROPDOWN_DIFFICULTY)
+    diffField.performClick()
+    val firstDiff = Difficulty.values().first().name
+    composeRule.onNodeWithText(firstDiff).performClick()
+    diffField.assertTextContains(firstDiff)
+  }
+
+  @Test
+  fun selectLocations_button_updates_label_with_point_count_and_invokes_callback() {
```
Comment (nolanprad) [inline app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreenTest.kt]: Missing test for empty point list label
Add assertion before updating:
selectBtn.assertTextContains("Select Locations") // no count

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreenTest.kt
Comment (nolanprad) [inline app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddHuntFieldsScreenTest.kt]: This test class is thorough, clear, and realistically simulates user interaction, covering almost every input scenario.
The only major gaps are localization resilience and a few missing edge-case assertions, but the structure and methodology are excellent

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt (hunk 1 above)
Comment (BuseThai) [reply app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddHuntScreen.kt]: Toast-based is more convenient for error handling, the error is tested by errorMsg, No worry

File: app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreen.kt (hunk 4 above)
Comment (BuseThai) [reply app/src/main/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreen.kt]: not in this file because it's a call back helper function

File: app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreenTest.kt [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  @Test
+  fun initial_empty_showsTitle_cancel_andConfirmDisabled() {
+    var canceled = false
+    setContent(onCancel = { canceled = true })
+
+    // Title visible
+    composeRule.onNodeWithText("Select Hunt Points").assertExists()
```
[the remark this answers is not in the document]
Comment (BuseThai) [reply app/src/androidTest/java/com/swentseekr/seekr/ui/addhunt/AddPointsMapScreenTest.kt]: this is not a test tag it's a text
